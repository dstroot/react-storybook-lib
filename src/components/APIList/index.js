import React, { useEffect, useState } from "react";

// render list from JSON response to URL
// <APIList url="https://www.dallasopendata.com/resource/f29b-eiwd.json" />
const APIList = ({ url }) => {
  const [records, setRecords] = useState([]);

  // Async get data via API
  const getData = async () => {
    try {
      // get data
      const res = await fetch(url);

      // process response
      if (res.ok) {
        const json = await res.json();
        setRecords(json);
      } else {
        setError(new Error(res.status + " " + res.statusText));
      }
    } catch (err) {
      setError(
        new Error(
          "The events API did not return any data. Check your privacy tools (such as Privacy Badger) and ad blockers (such as uBlock Origin)."
        )
      );
    }
  };

  // pass empty array to useEffect so we only
  // process once when mounting the component.
  useEffect(() => {
    getData();
  }, []); // empty array

  return (
    <div>
      <h1>Open Data List</h1>
      <NoKeyList records={records} />
    </div>
  );
};

// render list with keys
const List = ({ records }) => (
  <ul>
    {records.map(record => (
      <li key={record.docid}>
        Activity: {record.activity}, Amount: {record.chksubtot}, Year:{" "}
        {record.fy}, Vendor: {record.vendor}
      </li>
    ))}
    ;
  </ul>
);

// If there is no record id (key) you can use the index
// as the key but this is not recommended.
const NoKeyList = ({ records }) => (
  <ul>
    {records.map((record, index) => (
      <li key={index}>
        Department: {record.department}, Activity: {record.activity}, Amount:
        {record.chksubtot}, Year: {record.fy}, Vendor: {record.vendor}
      </li>
    ))}
  </ul>
);

export default APIList;
