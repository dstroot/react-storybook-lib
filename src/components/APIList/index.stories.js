import React from 'react';
import { storiesOf } from '@storybook/react';

// components
import APIList from '../APIList';

// APIList section
storiesOf('APIList', module).add('show APIList', () => (
  <APIList url="https://www.dallasopendata.com/resource/f29b-eiwd.json?$limit=20" />
));
