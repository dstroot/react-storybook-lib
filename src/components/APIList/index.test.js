import React from 'react';
import renderer from 'react-test-renderer';
import APIList from '../APIList';

describe('APIList', () => {
  it('it should render', () => {
    const component = renderer.create(
      <APIList url="https://www.dallasopendata.com/resource/f29b-eiwd.json?$limit=20" />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
