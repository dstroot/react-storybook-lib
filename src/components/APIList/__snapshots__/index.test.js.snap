// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`APIList it should render 1`] = `
<div>
  <h1>
    Open Data List
  </h1>
  <ul />
</div>
`;
